storage/mediator.go

```diff
@@ -199,3 +219,33 @@ func (m *mediator) reportLoop() {
 		}
 	}
 }
+
+func (m *mediator) databaseBootstrapStates() (databaseBootstrapStates, error) {
```

Comment:
hm could you migrate this retrieval to the `database/namspace/shard` types. Will reduce allocs (all the extra slices for namespace/shard each during call)

Reply:
I'm not sure this is a good idea. This is the only place in the codebase where these two actions (tick and flush) are in the same spot. If I push these checks further down I'm gonna have to do some incredibly weird contortions to get the measurements before the tick and then pass them to the flushing code.

The allocs are super small. In production, ticks take 3-10 minutes so we're talking about a slice with 2-8 entries (namespaces) and a map with 1024 non-interface types every few minutes.

Reply:
The retrieval should definitely be done in the `mediator.Tick()`, the nit was about the databaseBootstrapStates retrieval. i.e. you could do this instead - 

``` 
// types.go
type database interface { 
   BootstrapState() (databaseBootstrapState, error)
}

type databaseNamespace interface { 
   BootstrapState() (shardBootstrapStates, error)
}

type databaseShard interface { 
   BootstrapState() bootstrapStates
}

// database.go
func (d *db) BootstrapState() (...) {
	var (result ... , err error)
	d.RLock()
	defer d.RUnlock()
	for _, n := range d.namespaces.Iter() {
		result[n.Value().ID().String()], err = n.Value().BoostrapState()
		if err != nil {
		    return nil, err
		}
	}
}

// and so on
```

it'd avoid some un-necessary allocs you're currently doing: GetOwnedNamespaces()/GetOwnedShards etc

Reply:
Looks like I'll still have to allocated the slice for GetOwnedShards for the sake of iteration, can avoid the namespaces one though I think

Reply:
ok done